### ISSUE TITLE
Authorization in nested resource based on parent in index method

### ISSUE THREAD

**Issue by Obi-TOB** [CanCanCommunity/cancancan#365]
Hi,

 I can't get my head around the following:

I have a User model with specific roles (admin, privileged_user, normal_user).

The user has a has many relationship to posts.

I want to make different routes available for different user types (users and priviledged users) both having access to projects.

```
resources :users do
  resources :posts

resources :privileged_users do
  resources :posts
```
Note that I have only a User model (and roles), I don't have a PriviledgedUser model

Now in the `PostsController` I want to ensure that the priviledged_user has only access to it's own posts, i.e. I want to do authorisation in a nested route for the index action of the parent.
Basically I want to do:

```
def index
    @priv_user = User.find(:priviledged_user_id)
    authorize! :index, Post
    [...]
```
and in `class Ability`
```
if user.priviledged?
    can [:index], Post if user.id = @priv_user.id
```

However @priv_user is not accessible in the Ability class.
And as I'm in the index action I cannot use a block in the ability neither.

Any idea how to solve this?

**Comment by Obi-TOB** [comment]
For anyone else interested.

I used now in the post model:
```
load_and_authorize_resource :user, id_param: "privileged_user_id"
load_and_authorize_resource :user
```
and in `class Ability`
```
if user.priviledged?
    can [:show], User, :id =>user.id
```
I think that should do it. The logic is:
Try to instantiate a user by `privileged_user_id` parameter and autorize it.
If this is not possible try to instantiate a user by it normal `user_id` parameter and authorize it.

I don't like it very much, as I find it weird, that authorisation in the first case passes even if no user was identified (if no priviledged_user_id parameter is present), my intuitive feeling would have told me that it should stop.

So I still have to be careful when loading the posts itself, in particular as the relation is a bit more complicated (polymorphism). But here I will just follow https://github.com/CanCanCommunity/cancancan/wiki/Nested-Resources#polymorphic-associations

Any comments on why the above should not be done or any ideas for better ways are welcome... I was thinking about
- Subclassing the User
- Improve the unorthodox duplicate nested routes maybe use something like /user/:id/[role]/fb_posts.
Maybe this is not the right place to ask, but: Any recommendation on how role specific routes are implemented normally?

**Comment by coorasse** [comment]
Please refer to [Stackoverflow for questions](http://stackoverflow.com/questions/ask?tags=cancancan). Thanks.
